### PR TITLE
Flink SQL engine supports run across versions

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

To make sure flink sql engine build against flink-1.15 can run on flink-1.14.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: .github/workflows/master.yml [hunk 1]
```diff
@@ -128,11 +127,21 @@ jobs:
...
           ./build/mvn ${MVN_OPT} -pl ${TEST_MODULES} -Pflink-${{ matrix.flink }} ${{ matrix.flink-archive }} -am  clean install -DskipTests
+      - name: Test Flink
+        if: matrix.flink-archive == ''
+        run: |
           ./build/mvn ${MVN_OPT} -pl ${TEST_MODULES} -Pflink-${{ matrix.flink }} ${{ matrix.flink-archive }} test
+      - name: Cross-version test Flink
+        if: matrix.flink-archive != ''
+        run: |
+          IT_FLINK=`echo "${{ matrix.flink-archive }}" | grep -E 'flink\-([0-9]+\.[0-9]+.[0-9]+)\-bin' -o | grep -E '[0-9]+\.[0-9]+' -o`
+          ./build/mvn ${MVN_OPT} -pl ${IT_MODULE} -Pflink-${IT_FLINK} ${{ matrix.flink-archive }} test
```
Comment (pan3793) [inline .github/workflows/master.yml:144]: The cross-version verification composes of:
- flink-sql-engine compiled against flink-1.15
- flink-1.14 binary tgz
- flink-1.14 mini cluster
- flink-1.14 it test suites

Comment (pan3793) [reply .github/workflows/master.yml:144]: cc @deadwind4 FYI

**Comment by pan3793** [comment]
cc @link3280

**Comment by pan3793** [comment]
Also cc @benjobs

**Comment by pan3793** [comment]
Also cc @wolfboys

**Comment by link3280** [comment]
@pan3793 The newly introduced reflection utils seem very promising! I would take a detailed look later today. 👍🏼

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala [hunk 2]
```diff
@@ -116,17 +118,16 @@ class ExecuteStatement(
...
-                      rows ++= result.asInstanceOf[List[Row]]
-                    } else {
-                      // for flink 1.15+
-                      rows ++= result.map(r => convertToRow(r.asInstanceOf[RowData], dataTypes))
-                    }
+                val retrieveResultPage = DynMethods.builder("retrieveResultPage")
+                  .impl(executor.getClass, classOf[String], classOf[Int])
+                  .build(executor)
+                val _page = Integer.valueOf(page)
+                if (FlinkEngineUtils.isFlinkVersionEqualTo("1.14")) {
```
Comment (SteNicholas) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/operation/ExecuteStatement.scala:125]: It‘s better to use the `FlinkVersion` enum here.

**Review by SteNicholas — CHANGES_REQUESTED** [review]
Thanks for @pan3793 contribution. I left the minor comments for this pull request. PTAL.

File: externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FlinkEngineUtils.scala [hunk 3]
```diff
@@ -46,11 +46,20 @@ object FlinkEngineUtils extends Logging {
         logger.info(s"The current Flink version is $flinkVersion")
       case _ =>
         throw new UnsupportedOperationException(
-          s"The current Flink version is $flinkVersion, " +
-            s"Only Flink 1.14.x and 1.15 are supported, not supported in other versions")
+          s"You are using unsupported Flink version $flinkVersion, " +
+            s"only Flink 1.14 and 1.15 are supported now.")
```
Comment (SteNicholas) [inline externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FlinkEngineUtils.scala:50]: Please maintain a supported Flink version list here.

Comment (pan3793) [reply externals/kyuubi-flink-sql-engine/src/main/scala/org/apache/kyuubi/engine/flink/FlinkEngineUtils.scala]: good suggestion, updated

**Review by SteNicholas — APPROVED** [review]
LGTM.

File: .github/workflows/master.yml [hunk 4]
```diff
@@ -111,12 +111,11 @@ jobs:
...
-        # include:
-          # - java: 8
-            # flink: '1.15'
-            # flink-archive: '-Dflink.archive.mirror=https://archive.apache.org/dist/flink/flink-1.14.5 -Dflink.archive.name=flink-1.14.5-bin-scala_2.12.tgz'
-            # comment: 'verify-flink-1.14'
+        include:
+          - java: 8
+            flink: '1.15'
+            flink-archive: '-Dflink.archive.mirror=https://archive.apache.org/dist/flink/flink-1.14.5 -Dflink.archive.name=flink-1.14.5-bin-scala_2.12.tgz'
+            comment: 'verify-flink-1.14'
```
Comment (deadwind4) [inline .github/workflows/master.yml:118]:
```suggestion
            comment: 'verify-engine-1.15-on-flink-1.14 '
```

Comment (pan3793) [reply .github/workflows/master.yml]: changed to `verify-on-flink-1.14-binary`, 1.15 is acutally the default flink version, we add this verification to ensure the default released binary tgz works on all supported versions of flink binary.

**Comment by deadwind4** [comment]
LGTM.

**Review by deadwind4 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks all for review, merging to master

**Pull request closed without merging**